From 11fadf1351009e3f8ebad6c6aa7cfe7536cf808e Mon Sep 17 00:00:00 2001
From: Yi-Chia Hsieh <Yi-Chia.Hsieh@mediatek.com>
Date: Thu, 24 Jul 2025 23:00:55 +0000
Subject: [PATCH 212/254] mtk: hostapd: add support for broadcast TWT

Add support for broadcast TWT
1. Add add/del/get command for broadcast TWT
2. Maintain a list of broadcast TWT service period
3. Send nl80211 command to cfg80211 for creating broadcast TWT service period

Signed-off-by: Yi-Chia Hsieh <Yi-Chia.Hsieh@mediatek.com>
---
 hostapd/ctrl_iface.c               | 184 +++++++++++++++++++++++++++++
 hostapd/hostapd_cli.c              |  27 +++++
 src/ap/ap_drv_ops.h                |  17 +++
 src/ap/hostapd.c                   |   5 +
 src/ap/hostapd.h                   |  25 ++++
 src/drivers/driver.h               |  25 ++++
 src/drivers/driver_nl80211.c       |  77 ++++++++++++
 src/drivers/driver_nl80211_event.c |   2 +
 8 files changed, 362 insertions(+)

diff --git a/hostapd/ctrl_iface.c b/hostapd/ctrl_iface.c
index 2114b4684..cb00537db 100644
--- a/hostapd/ctrl_iface.c
+++ b/hostapd/ctrl_iface.c
@@ -3141,6 +3141,182 @@ static int hostapd_ctrl_iface_color_change(struct hostapd_iface *iface,
 	return -1;
 #endif /* NEED_AP_MLME */
 }
+
+static int
+hostapd_ctrl_iface_add_broadcast_twt(struct hostapd_data *hapd, char *cmd,
+				     char *buf, size_t buflen)
+{
+	char *pos = cmd, *end;
+	struct broadcast_twt_param *param;
+
+	param = os_malloc(sizeof(*param));
+	if (param == NULL)
+		return -ENOBUFS;
+
+	param->id = strtol(pos, &end, 10);
+	if (pos == end) {
+		wpa_printf(MSG_ERROR, "broadcast_twt: invalid btwt id");
+		goto fail;
+	}
+	pos = end;
+
+	param->min_wake_dur = strtol(pos, &end, 10);
+	if (pos == end) {
+		wpa_printf(MSG_ERROR, "broadcast_twt: invalid min_wake_dur");
+		goto fail;
+	}
+	pos = end;
+
+	param->wake_intv_mantissa = strtol(pos, &end, 10);
+	if (pos == end) {
+		wpa_printf(MSG_ERROR, "broadcast_twt: invalid wake_intv_mantissa");
+		goto fail;
+	}
+	pos = end;
+
+	param->wake_intv_exp = strtol(pos, &end, 10);
+	if (pos == end) {
+		wpa_printf(MSG_ERROR, "broadcast_twt: invalid wake_intv_exp");
+		goto fail;
+	}
+	pos = end;
+
+	param->trigger = 0;
+	param->flow_type = 0;
+	param->recommendation = 0;
+	param->persist = 255;
+
+	pos = os_strstr(cmd, "trigger=");
+	if (pos)
+		param->trigger = !!atoi(pos + 8);
+
+	pos = os_strstr(cmd, "flow_type=");
+	if (pos)
+		param->flow_type = !!atoi(pos + 10);
+
+	pos = os_strstr(cmd, "recommendation=");
+	if (pos)
+		param->recommendation = atoi(pos + 15);
+
+	pos = os_strstr(cmd, "persist=");
+	if (pos)
+		param->persist = atoi(pos + 8);
+
+	if (param->id > 31 ||
+	    param->wake_intv_exp > 31 ||
+	    param->recommendation > 7) {
+		wpa_printf(MSG_ERROR, "broadcast_twt: invalid argument");
+		goto fail;
+	}
+
+	if (hostapd_drv_add_btwt(hapd, param)) {
+		wpa_printf(MSG_ERROR, "broadcast_twt: driver add fail");
+		goto fail;
+	}
+
+	hapd->broadcast_twt_cnt++;
+	dl_list_add_tail(&hapd->btwt_sp_list, &param->list);
+	ieee802_11_set_beacon_per_bss_only(hapd);
+
+	return os_snprintf(buf, buflen, "broadcast_twt: id=%hhu min_wake_dur=%hhu "
+			   "wake_intv_mantissa=%hu wake_intv_exp=%hu "
+			   "trigger=%hhu flow_type=%hhu "
+			   "recommendation=%hhu persist=%hhu in bssid=" MACSTR "\n",
+			   param->id, param->min_wake_dur, param->wake_intv_mantissa,
+			   param->wake_intv_exp, param->trigger, param->flow_type,
+			   param->recommendation, param->persist,
+			   MAC2STR(hapd->own_addr));
+
+fail:
+	os_free(param);
+	return -1;
+}
+
+static int
+hostapd_ctrl_iface_del_broadcast_twt(struct hostapd_data *hapd, char *cmd,
+				     char *buf, size_t buflen)
+{
+	char *pos = cmd, *end;
+	bool found = false;
+	int ret;
+	struct broadcast_twt_param *param, *iter;
+	u8 id;
+
+	if (hapd->broadcast_twt_cnt == 0) {
+		wpa_printf(MSG_ERROR, "broadcast_twt: nothing to be deleted");
+		return -1;
+	}
+
+	id = strtol(pos, &end, 10);
+	if (pos == end) {
+		wpa_printf(MSG_ERROR, "broadcast_twt: invalid id provided");
+		return -1;
+	}
+	pos = end;
+
+	if (hostapd_drv_del_btwt(hapd, id)) {
+		wpa_printf(MSG_ERROR, "broadcast_twt: driver delete fail");
+		return -1;
+	}
+
+	dl_list_for_each_safe(param, iter, &hapd->btwt_sp_list,
+			      struct broadcast_twt_param, list) {
+		if (param->id == id) {
+			found = true;
+			break;
+		}
+	}
+
+	if (!found) {
+		wpa_printf(MSG_ERROR, "broadcast_twt: invalid id");
+		return -1;
+	}
+
+	hapd->broadcast_twt_cnt--;
+	dl_list_del(&param->list);
+	ieee802_11_set_beacon_per_bss_only(hapd);
+	ret = os_snprintf(buf, buflen, "broadcast_twt: id=%hhu in bssid="
+			  MACSTR " deleted\n", param->id,
+			  MAC2STR(hapd->own_addr));
+	os_free(param);
+	return ret;
+}
+
+static int
+hostapd_ctrl_iface_get_broadcast_twt(struct hostapd_data *hapd, char *buf, size_t buflen)
+{
+	int ret;
+	char *pos, *end;
+	struct broadcast_twt_param *param;
+
+	pos = buf;
+	end = buf + buflen;
+	ret = os_snprintf(pos, end - pos,
+			  "broadcast_twt: %u set(s) of broadcast twt\n",
+			  hapd->broadcast_twt_cnt);
+	if (os_snprintf_error(end - pos, ret))
+		return pos - buf;
+	pos += ret;
+
+	dl_list_for_each(param, &hapd->btwt_sp_list,
+			 struct broadcast_twt_param, list) {
+		ret = os_snprintf(pos, end - pos, "id=%hhu min_wake_dur=%hhu "
+				  "wake_intv_mantissa=%hu wake_intv_exp=%hhu "
+				  "trigger=%hhu flow_type=%hhu "
+				  "recommendation=%hhu persist=%hhu in bssid=" MACSTR "\n",
+				  param->id, param->min_wake_dur,
+				  param->wake_intv_mantissa, param->wake_intv_exp,
+				  param->trigger, param->flow_type,
+				  param->recommendation, param->persist,
+				  MAC2STR(hapd->own_addr));
+
+		if (os_snprintf_error(end - pos, ret))
+			return pos - buf;
+		pos += ret;
+	}
+
+	return pos - buf;
+}
 #endif /* CONFIG_IEEE80211AX */
 
 
@@ -6736,6 +6912,14 @@ static int hostapd_ctrl_iface_receive_process(struct hostapd_data *hapd,
 	} else if (os_strncmp(buf, "COLOR_CHANGE ", 13) == 0) {
 		if (hostapd_ctrl_iface_color_change(hapd->iface, buf + 13))
 			reply_len = -1;
+	} else if (os_strncmp(buf, "ADD_BROADCAST_TWT ", 18) == 0) {
+		reply_len = hostapd_ctrl_iface_add_broadcast_twt(hapd, buf + 18,
+				reply, reply_size);
+	} else if (os_strncmp(buf, "DEL_BROADCAST_TWT ", 18) == 0) {
+		reply_len = hostapd_ctrl_iface_del_broadcast_twt(hapd, buf + 18,
+				reply, reply_size);
+	} else if (os_strncmp(buf, "GET_BROADCAST_TWT", 17) == 0) {
+		reply_len = hostapd_ctrl_iface_get_broadcast_twt(hapd, reply, reply_size);
 #endif /* CONFIG_IEEE80211AX */
 	} else if (os_strncmp(buf, "NOTIFY_CW_CHANGE ", 17) == 0) {
 		if (hostapd_ctrl_iface_notify_cw_change(hapd, buf + 17))
diff --git a/hostapd/hostapd_cli.c b/hostapd/hostapd_cli.c
index faf141dcf..cb2fb9b68 100644
--- a/hostapd/hostapd_cli.c
+++ b/hostapd/hostapd_cli.c
@@ -1896,6 +1896,24 @@ hostapd_cli_cmd_clear_disconn_counter(struct wpa_ctrl *ctrl, int argc, char *arg
 	return wpa_ctrl_command(ctrl, "CLEAR_DISCONN_COUNTER");
 }
 
+static int hostapd_cli_cmd_add_broadcast_twt(struct wpa_ctrl *ctrl, int argc,
+					     char *argv[])
+{
+	return hostapd_cli_cmd(ctrl, "ADD_BROADCAST_TWT", 4, argc, argv);
+}
+
+static int hostapd_cli_cmd_del_broadcast_twt(struct wpa_ctrl *ctrl, int argc,
+					     char *argv[])
+{
+	return hostapd_cli_cmd(ctrl, "DEL_BROADCAST_TWT", 1, argc, argv);
+}
+
+static int hostapd_cli_cmd_get_broadcast_twt(struct wpa_ctrl *ctrl, int argc,
+					     char *argv[])
+{
+	return hostapd_cli_cmd(ctrl, "GET_BROADCAST_TWT", 0, 0, NULL);
+}
+
 struct hostapd_cli_cmd {
 	const char *cmd;
 	int (*handler)(struct wpa_ctrl *ctrl, int argc, char *argv[]);
@@ -1997,6 +2015,15 @@ static const struct hostapd_cli_cmd hostapd_cli_commands[] = {
 	{ "color_change", hostapd_cli_cmd_color_change, NULL,
 	  "<color> = initiate BSS color change to set the specified color\n"
 	  "Value 0 will disable the color.\n"},
+	{ "add_btwt", hostapd_cli_cmd_add_broadcast_twt, NULL,
+	  "<id> <min_wake_dur> <wake_intv_manti> <wake_intv_exp>\n"
+	  "  [trigger=<0/1>] [flow_type=<0/1>]\n"
+	  "  [recommendation=0~7] [persist=0~255]\n"
+	  "  = Add Broadcast TWT" },
+	{ "del_btwt", hostapd_cli_cmd_del_broadcast_twt, NULL,
+		" <id> = Delete Broadcast TWT"},
+	{ "get_btwt", hostapd_cli_cmd_get_broadcast_twt, NULL,
+		" = Show existing Broadcast TWT"},
 #endif /* CONFIG_IEEE80211AX */
 	{ "notify_cw_change", hostapd_cli_cmd_notify_cw_change, NULL,
 	  "<channel_width> = 0 - 20 MHz, 1 - 40 MHz, 2 - 80 MHz, 3 - 160 MHz" },
diff --git a/src/ap/ap_drv_ops.h b/src/ap/ap_drv_ops.h
index bea7c4183..fbcf9532d 100644
--- a/src/ap/ap_drv_ops.h
+++ b/src/ap/ap_drv_ops.h
@@ -393,6 +393,23 @@ static inline int hostapd_drv_switch_color(struct hostapd_data *hapd,
 
 	return hapd->driver->switch_color(hapd->drv_priv, settings);
 }
+
+static inline int hostapd_drv_add_btwt(struct hostapd_data *hapd,
+				       struct broadcast_twt_param *param)
+{
+	if (!hapd->driver || !hapd->driver->add_btwt || !hapd->drv_priv)
+		return -1;
+
+	return hapd->driver->add_btwt(hapd->drv_priv, param);
+}
+
+static inline int hostapd_drv_del_btwt(struct hostapd_data *hapd, u8 id)
+{
+	if (!hapd->driver || !hapd->driver->del_btwt || !hapd->drv_priv)
+		return -1;
+
+	return hapd->driver->del_btwt(hapd->drv_priv, id);
+}
 #endif /* CONFIG_IEEE80211AX */
 
 static inline int hostapd_drv_status(struct hostapd_data *hapd, char *buf,
diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index b0e068241..06c17e5fc 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -1914,6 +1914,11 @@ setup_mld:
 		}
 	}
 
+	if (hostapd_twt_init(hapd)) {
+		wpa_printf(MSG_ERROR, "TWT initialization failed.");
+		return -1;
+	}
+
 	if (!hostapd_drv_none(hapd) && vlan_init(hapd)) {
 		wpa_printf(MSG_ERROR, "VLAN initialization failed.");
 		return -1;
diff --git a/src/ap/hostapd.h b/src/ap/hostapd.h
index 2921aa38c..af4d88afb 100644
--- a/src/ap/hostapd.h
+++ b/src/ap/hostapd.h
@@ -391,6 +391,9 @@ struct hostapd_data {
 	struct os_reltime first_color_collision;
 	struct os_reltime last_color_collision;
 	u64 color_collision_bitmap;
+
+	struct dl_list btwt_sp_list; /* list of struct broadcast_twt_param */
+	u8 broadcast_twt_cnt;
 #endif /* CONFIG_IEEE80211AX */
 
 #ifdef CONFIG_P2P
@@ -1114,4 +1117,26 @@ static inline bool ap_pmf_enabled(struct hostapd_bss_config *conf)
 		conf->rsn_override_mfp_2 != NO_MGMT_FRAME_PROTECTION;
 }
 
+static inline u32 hostapd_get_radio_mask(struct hostapd_data *hapd)
+{
+#ifdef CONFIG_IEEE80211BE
+	if (hapd->iface->current_hw_info) {
+		if (hapd->conf->mld_ap)
+			return hapd->iface->radio_mask;
+		else
+			return 1 << hapd->iface->current_hw_info->hw_idx;
+	}
+#endif /* CONFIG_IEEE80211BE */
+	return 0;
+}
+
+static inline int hostapd_twt_init(struct hostapd_data *hapd)
+{
+#ifdef CONFIG_IEEE80211AX
+	dl_list_init(&hapd->btwt_sp_list);
+	hapd->broadcast_twt_cnt = 0;
+#endif /* CONFIG_IEEE80211AX */
+	return 0;
+}
+
 #endif /* HOSTAPD_H */
diff --git a/src/drivers/driver.h b/src/drivers/driver.h
index b3339ba3f..ff99eb323 100644
--- a/src/drivers/driver.h
+++ b/src/drivers/driver.h
@@ -2960,6 +2960,21 @@ struct cca_settings {
 	int link_id;
 };
 
+/**
+ * struct broadcast_twt_param - Parameters for broadcast TWT service period
+ */
+struct broadcast_twt_param {
+	struct dl_list list;
+	u8 id;
+	u8 min_wake_dur;
+	u16 wake_intv_mantissa;
+	u8 wake_intv_exp;
+	bool trigger;
+	bool flow_type;
+	u8 recommendation;
+	u8 persist;
+};
+
 /* TDLS peer capabilities for send_tdls_mgmt() */
 enum tdls_peer_capability {
 	TDLS_PEER_HT = BIT(0),
@@ -4822,6 +4837,16 @@ struct wpa_driver_ops {
 	 */
 	int (*switch_color)(void *priv, struct cca_settings *settings);
 
+	/**
+	* add_btwt - add new broadcast twt service period
+	*/
+	int (*add_btwt)(void *priv, struct broadcast_twt_param *param);
+
+	/**
+	* del_btwt - delete broadcast twt service period
+	*/
+	int (*del_btwt)(void *priv, u8 id);
+
 	/**
 	 * add_tx_ts - Add traffic stream
 	 * @priv: Private driver interface data
diff --git a/src/drivers/driver_nl80211.c b/src/drivers/driver_nl80211.c
index 31f8239e7..ec7a31009 100644
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -12307,6 +12307,81 @@ error:
 	wpa_printf(MSG_DEBUG, "nl80211: Could not build color switch request");
 	return ret;
 }
+
+int nl80211_add_btwt(void *priv, struct broadcast_twt_param *param)
+{
+	struct i802_bss *bss = priv;
+	struct wpa_driver_nl80211_data *drv = bss->drv;
+	struct nl_msg *msg;
+	struct nlattr *data;
+	int ret = -ENOBUFS;
+
+	wpa_printf(MSG_DEBUG, "nl80211: add Broadcast TWT, ifname=%s", bss->ifname);
+	msg = nl80211_cmd_msg(bss, 0, NL80211_CMD_ADD_BROADCAST_TWT);
+	if (!msg)
+		goto fail;
+
+	data = nla_nest_start(msg, NL80211_ATTR_BROADCAST_TWT_PARAMS);
+	if (!data)
+		goto fail;
+
+	if (nla_put_u8(msg, NL80211_BROADCAST_TWT_ATTR_ID, param->id) ||
+	    nla_put_u8(msg, NL80211_BROADCAST_TWT_ATTR_MIN_WAKE_DUR, param->min_wake_dur) ||
+	    nla_put_u16(msg, NL80211_BROADCAST_TWT_ATTR_MANTISSA, param->wake_intv_mantissa) ||
+	    nla_put_u8(msg, NL80211_BROADCAST_TWT_ATTR_EXPONENT, param->wake_intv_exp) ||
+	    nla_put_u8(msg, NL80211_BROADCAST_TWT_ATTR_RECOMMENDATION, param->recommendation) ||
+	    nla_put_u8(msg, NL80211_BROADCAST_TWT_ATTR_PERSISTENCE, param->persist) ||
+	    (param->trigger && nla_put_flag(msg, NL80211_BROADCAST_TWT_ATTR_TRIGGER)) ||
+	    (param->flow_type && nla_put_flag(msg, NL80211_BROADCAST_TWT_ATTR_FLOW_TYPE)))
+		goto fail;
+
+	nla_nest_end(msg, data);
+
+	ret = send_and_recv_cmd(drv, msg);
+	if (ret)
+		wpa_printf(MSG_DEBUG, "Failed to add Broadcast TWT. ret=%d (%s)",
+			   ret, strerror(-ret));
+	return ret;
+
+fail:
+	nlmsg_free(msg);
+	wpa_printf(MSG_DEBUG, "nl80211: Could not build add broadcast TWT request");
+	return ret;
+}
+
+int nl80211_del_btwt(void *priv, u8 id)
+{
+	struct i802_bss *bss = priv;
+	struct wpa_driver_nl80211_data *drv = bss->drv;
+	struct nl_msg *msg;
+	struct nlattr *data;
+	int ret = -ENOBUFS;
+
+	wpa_printf(MSG_DEBUG, "nl80211: del Broadcast TWT, ifname=%s", bss->ifname);
+	msg = nl80211_bss_msg(bss, 0, NL80211_CMD_DEL_BROADCAST_TWT);
+	if (!msg)
+		goto fail;
+
+	data = nla_nest_start(msg, NL80211_ATTR_BROADCAST_TWT_PARAMS);
+	if (!data)
+		goto fail;
+
+	if (nla_put_u8(msg, NL80211_BROADCAST_TWT_ATTR_ID, id))
+		goto fail;
+
+	nla_nest_end(msg, data);
+
+	ret = send_and_recv_cmd(drv, msg);
+	if (ret)
+		wpa_printf(MSG_DEBUG, "Failed to del Broadcast TWT. ret=%d (%s)",
+			   ret, strerror(-ret));
+	return ret;
+
+fail:
+	nlmsg_free(msg);
+	wpa_printf(MSG_DEBUG, "nl80211: Could not build del broadcast TWT request");
+	return ret;
+}
 #endif /* CONFIG_IEEE80211AX */
 
 
@@ -17109,6 +17184,8 @@ const struct wpa_driver_ops wpa_driver_nl80211_ops = {
 	.switch_channel = nl80211_switch_channel,
 #ifdef CONFIG_IEEE80211AX
 	.switch_color = nl80211_switch_color,
+	.add_btwt = nl80211_add_btwt,
+	.del_btwt = nl80211_del_btwt,
 #endif /* CONFIG_IEEE80211AX */
 #ifdef ANDROID_P2P
 	.set_noa = wpa_driver_set_p2p_noa,
diff --git a/src/drivers/driver_nl80211_event.c b/src/drivers/driver_nl80211_event.c
index 0e94d1e51..2a0b08c8e 100644
--- a/src/drivers/driver_nl80211_event.c
+++ b/src/drivers/driver_nl80211_event.c
@@ -194,6 +194,8 @@ static const char * nl80211_command_to_string(enum nl80211_commands cmd)
 	C2S(NL80211_CMD_SET_ATTLM)
 	C2S(NL80211_CMD_TSF_OFFSET_EVENT)
 	C2S(NL80211_CMD_NOTIFY_CRIT_UPDATE)
+	C2S(NL80211_CMD_ADD_BROADCAST_TWT)
+	C2S(NL80211_CMD_DEL_BROADCAST_TWT)
 	C2S(__NL80211_CMD_AFTER_LAST)
 	}
 #undef C2S
-- 
2.45.2

